src/lib/core/App.tsx

```diff
@@ -0,0 +1,43 @@
+import * as React from 'react'
+import { History, Location } from 'history'
+import { Router, RouteConfig } from './Router'
+
+export interface AppProps {
+  routes: RouteConfig[]
+  history: History
+}
+
+export interface AppState {
+  location: Location
+}
+
+export interface TransitionEnterEvent {
+  incomingLocation: Location
+  outgoingLocation: Location
+  redirect: RedirectFunction
+}
+
+export type RedirectFunction = (l: Location) => void
+
+/**
+ * Root application container.
+ */
+export class App extends React.PureComponent<AppProps> {
+  router = new Router(this.props.routes, this.props.history)
+
+  state: AppState = {
+    location: this.props.history.location,
+  }
+
+  render() {
+    const { route } = this.router
+
+    if (route) {
+      const Handler = route.handler
+      return <Handler {...route} />
+
+    } else {
+      return null
```

Comment:
What's the result of returning null? Does the user see a 404?

Reply:
No, it just renders an empty page. Was thinking that 404 should be handled with a route matching '*'

Reply:
Could argue that it should throw an error here instead, though. What do you think?

Reply:
I think an error would probably be better - from my reading of the code it looks like an error state. The error would be more descriptive as well...